lib/puppet/util/at_fork/solaris.rb

```diff
@@ -0,0 +1,155 @@
+# contract is terminated.
+class Puppet::Util::AtFork::Solaris < Puppet::Util::AtFork
+  private
+
+  {
+    'libcontract.so.1' => [
+      #function name,            return value type, parameter types, ...
+      [:ct_ctl_abandon,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_activate,        Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_clear,           Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+
+      [:ct_tmpl_set_informative, Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_tmpl_set_critical,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_param,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_fatal,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+
+      [:ct_status_read,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT, Fiddle::TYPE_VOIDP],
+
+      [:ct_status_get_id,        Fiddle::TYPE_INT,  Fiddle::TYPE_VOIDP],
+      [:ct_status_free,          Fiddle::TYPE_VOID, Fiddle::TYPE_VOIDP],
+    ],
+  }.each do |library, functions|
+    libhandle = Fiddle::Handle.new(library)
+
+    functions.each do |f|
+      define_method f[0], Fiddle::Function.new(libhandle[f[0].to_s], f[2..-1], f[1]).method(:call).to_proc
+    end
+  end
+
+  CTFS_PR_ROOT = File.join('', %w(system contract process))
+  CTFS_PR_TEMPLATE = File.join(CTFS_PR_ROOT, %q(template))
+  CTFS_PR_LATEST = File.join(CTFS_PR_ROOT, %q(latest))
+
+  CT_PR_PGRPONLY = 0x4
+  CT_PR_EV_HWERR = 0x20
+
+  CTD_COMMON = 0
+
+  def activate_new_contract_template()
+    begin
+      tmpl = File.open(CTFS_PR_TEMPLATE, File::RDWR)
+    rescue => detail
+      Puppet.log_exception(detail, _('Failed to open a new process contract template'))
+      return
+    end
+
+    begin
+      tmpl_fd = tmpl.fileno
+
+      raise if ct_pr_tmpl_set_param(tmpl_fd, CT_PR_PGRPONLY) != 0
+      raise if ct_pr_tmpl_set_fatal(tmpl_fd, CT_PR_EV_HWERR) != 0
+      raise if ct_tmpl_set_critical(tmpl_fd, 0) != 0
+      raise if ct_tmpl_set_informative(tmpl_fd, CT_PR_EV_HWERR) != 0
+
+      raise if ct_tmpl_activate(tmpl_fd) != 0
+    rescue
+      tmpl.close()
+      Puppet.err(_('Failed to configure / activate a new process contract template'))
+      return
+    end
+
+    @tmpl = tmpl
+  end
+
+  def deactivate_contract_template(where)
+    return if @tmpl.nil?
+
+    tmpl = @tmpl
+    @tmpl = nil
+
+    begin
+      raise if ct_tmpl_clear(tmpl.fileno) != 0
+    rescue
+      Puppet.fatal(_("Failed to deactivate process contract template in the #{where} process"))
+    ensure
+      tmpl.close()
+    end
+  end
+
+  def get_latest_child_contract_id()
+    begin
+      stat = File.open(CTFS_PR_LATEST, File::RDONLY)
+    rescue => detail
+      Puppet.log_exception(detail, _('Unable to access latest child process contract status'))
+      return nil
+    end
+
+    begin
+      stathdl = Fiddle::Pointer.new(0)
```

Comment:
Yeah, makes sense. I assumed that fiddle allocated native memory for the ptr itself, even when the address is 0x0, but it doesn't look like it behaves that way. I assumed it worked like FFI, which does allocate memory:

```ruby
static VALUE
memptr_initialize(int argc, VALUE* argv, VALUE self)
{
    VALUE size = Qnil, count = Qnil, clear = Qnil;
    int nargs = rb_scan_args(argc, argv, "12", &size, &count, &clear);

    memptr_malloc(self, rbffi_type_size(size), nargs > 1 ? NUM2LONG(count) : 1,
        RTEST(clear) || clear == Qnil);
```

About the NULL business, it might be clearer to do `stathdl = Fiddle::NULL`